### PR TITLE
small improvement on can-assign which checks that the property exists …

### PR DESCRIPTION
**Pull request by cherifGsoul**
Small improvement on can-assign which checks that the property exists in the supplier object
@justinbmeyer @matthewp can you check if it works for can ecosystem

### CODE REVIEW

**Comment by cherifGsoul** [comment]
it breaks on `can-util/dom/attr: attributes event`

**Comment by matthewp** [comment]
I would test this against the native Object.assign and see if it does hasOwnProperty. If not then we shouldn't either.

**Comment by cherifGsoul** [comment]
@matthewp I think it does, the test breaks the same way as with `Object.assign` and the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign pollyfil example checks with `hasOwnProperty` also

File: js/assign/assign.js [hunk 1]
```diff
@@ -25,7 +25,9 @@
 
 module.exports = function (d, s) {
 	for (var prop in s) {
-		d[prop] = s[prop];
+		if (s.hasOwnProperty(prop)) {
```
Comment (justinbmeyer) [inline js/assign/assign.js]: `Object.prototype.hasOwnProperty.call(s, prop)`

**Comment by cherifGsoul** [comment]
I think It breaks here https://github.com/canjs/can-util/blob/master/dom/events/events.js#L33

**Comment by justinbmeyer** [comment]
@cherifGsoul Odd.  Why would event have not "owned" properties?

**Comment by cherifGsoul** [comment]
@justinbmeyer I did a debugging and I saw that the ```ev``` object and ```event``` dont share the same properties

**Comment by cherifGsoul** [comment]
To accept this PR I think we should have our version of ```extend``` like ```jQuery.extend``` to handle cases where ```var anObj = assign({}, otherObj)``` like this line https://github.com/canjs/can-util/blob/master/dom/events/attributes/attributes.js#L39 .
The actuel assign is ```extend``` more than ```Object.assign```

**Comment by justinbmeyer** [comment]
Something like:

```
var person = {first: "Justin", last: "Meyer"};
var user = {};
assign({},{...}) //-> {}
Object.assign({},{}) //-> {}
```

**Comment by cherifGsoul** [comment]
To use ```assign``` the reciever should have the same keys (props) as the supplier for example:
```
var person = {first: "Justin", last: "Meyer"};
var user = assign({}, person); => {} //empty object
```
dosent works because it tries to override only the same properties (we use ```hasOwnProperty```)

In contrast ```extend``` will copy all keys to the new object:
```
var person = {first: "Justin", last: "Meyer"};
var user = extend({}, person); => {first: "Justin", last: "Meyer"}
```
```
assign => Override
extend => inheritance 
```
[underscore.js](https://github.com/jashkenas/underscore/blob/master/underscore.js#L1087) [distinguish](https://github.com/jashkenas/underscore/blob/master/underscore.js#L1091) between the two behaviors

**Comment by justinbmeyer** [comment]
@cherifGsoul if this is true, why does the first example [on mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) example show:

```js
var obj = { a: 1 };
var copy = Object.assign({}, obj);
console.log(copy); // { a: 1 }
```

**Comment by cherifGsoul** [comment]
@justinbmeyer good point but the pollyfil shows other thing the it employs the ```hasOwnProperty``` function too

**Comment by justinbmeyer** [comment]
@cherifGsoul it also seems that the polyfill will copy all properties regardless if they are on the destination or not:

```js
if (typeof Object.assign != 'function') {
  Object.assign = function(target, varArgs) { // .length of function is 2
    'use strict';
    if (target == null) { // TypeError if undefined or null
      throw new TypeError('Cannot convert undefined or null to object');
    }

    var to = Object(target);

    for (var index = 1; index < arguments.length; index++) {
      var nextSource = arguments[index];

      if (nextSource != null) { // Skip over if undefined or null
        for (var nextKey in nextSource) {
          // Avoid bugs when hasOwnProperty is shadowed
          if (Object.prototype.hasOwnProperty.call(nextSource, nextKey)) {
            to[nextKey] = nextSource[nextKey];
          }
        }
      }
    }
    return to;
  };
}
```

**Comment by justinbmeyer** [comment]
> @justinbmeyer good point but the pollyfil shows other thing the it employs the hasOwnProperty function too

The `can-util/js/assign/` module should mimic the behavior of `Object.assign`.  If we need an `can-util/js/assign-all` to include non-enumerable properties, we can make it and use it where needed.

**Comment by cherifGsoul** [comment]
Im confused about the doc of MDN I will test all this

**Comment by cherifGsoul** [comment]
Not ready to merge

**Comment by cherifGsoul** [comment]
I think I did something wrong when trying to resolve conflicts !!

**Comment by cherifGsoul** [comment]
to be closed and working on new branch.

**Pull request closed without merging**